test/master-playlist-controller.test.js

```diff
@@ -1067,6 +1067,217 @@ function(assert) {
                'less than low water line');
 });
 
+QUnit.test('blacklists playlist on earlyabort', function(assert) {
+  this.masterPlaylistController.mediaSource.trigger('sourceopen');
+  // master
+  this.standardXHRResponse(this.requests.shift());
+  // media
+  this.standardXHRResponse(this.requests.shift());
+
+  let mediaChanges = [];
+  const playlistLoader = this.masterPlaylistController.masterPlaylistLoader_;
+  const currentMedia = playlistLoader.media();
+  const origMedia = playlistLoader.media.bind(playlistLoader);
+  const origWarn = videojs.log.warn;
+  let warnings = [];
+
+  this.masterPlaylistController.masterPlaylistLoader_.media = (media) => {
+    if (media) {
+      mediaChanges.push(media);
+    }
+    return origMedia(media);
+  };
+
+  videojs.log.warn = (text) => warnings.push(text);
+
+  assert.notOk(currentMedia.excludeUntil > 0, 'playlist not blacklisted');
+  assert.equal(mediaChanges.length, 0, 'no media change');
+
+  this.masterPlaylistController.mainSegmentLoader_.trigger('earlyabort');
+
+  assert.ok(currentMedia.excludeUntil > 0, 'playlist blacklisted');
+  assert.equal(mediaChanges.length, 0, 'no media change');
+  assert.equal(warnings.length, 1, 'one warning logged');
+  assert.equal(warnings[0],
+               'Problem encountered with the current HLS playlist. ' +
+                 'Aborted early because we don\'t have the bandwidth to complete the ' +
+                 'request without rebuffering.',
+               'warning message is correct');
+
+  videojs.log.warn = origWarn;
+});
+
+QUnit.test('does not get stuck in a loop due to inconsistent network/caching',
```

Comment:
This test looks like a bit of a bear.  It might be helpful to do one of a few things to make things clearer to future explorers:

1. Find a way to break this test up into a few smaller tests... maybe evaluate each stage as its own test.  I don't know if qunit lets your group things as you can in [mocha](https://mochajs.org/#asynchronous-code) with describe blocks, but you could have something like:

```
describe('caching', () => {
  describe('does not get stuck in a loop due to inconsistent network/caching' () => {
    before(() => { // bootstrap data these tests use });
    function setUpStep1() => { ... };
    function setUpStep2() => { ... };

    it('tests the first step', () => {
     setupStep1();
     assert(x,y, 'output of step 1 looks good!');
    });

   it ('tests second step', () => {
     setupStep1();
     setupStep2();
     assert(x,y, 'output of step 2 looks good!');
   });
  })
})
```

This might yield more code, but the tests themselves might end up looking a bit clearer, since each is testing a distinct scenario with regards to your use case.  It might make debugging a bit easier (instead of the caching test as a whole breaking on an assert that you need to dig through, the test that broke should give you the information that you need to debug a little bit faster).

There is a library that could work (there could be others), I don't know if we want to introduce it here: https://github.com/square/qunit-bdd

2. If you don't want to do that, maybe adding a comment block on this test that describes what's going on would be useful.  Looking at this test I'm aware that there is some kind of loop that is caused by caching, but the use case is hard to tell without reading through the entire test.  If someone introduces something that breaks this test it might be difficult to debug whether something is supposed to break based on the change, or if their change broke functionality that we need to preserve.

3. Maybe find a way to break some things up into separate files.  The path of having a test per file seems to have kind of outgrown its usefulness here.  We could break the per-file tests into folders, then have test files for various scenarios (or something).  Basically breaking this part of the file into something like test/master-playlist-controller/caching.test.js.